### ISSUE TITLE
Putting book inside their own folder under authors name

### ISSUE THREAD

**Issue by KatCountryGirl** [rmcrackan/Libation#644]
Hello first off let me say thank you I LOVE your program. 

I would love the option to rename files by author folder then by book in its own folder inside the author folder. Right now I can do this this author/book files but if there are multiple books they are all just in the author folder.  And save the image to that internal book folder.

I did try to add 

<author>
\<title> - <series> - <language short> [<id>]

Also the option to use the renamer for files I got outside of libation would be amazing as I have yet to find a decent program to rename my acquired  files like yours does. Thanks again!

**Comment by Mbucari** [comment]
@KatCountryGirl I'm glad you're enjoying it!

Have you tried <author>\<title short>
> I did try to add
> \<title> - - []

Github likes to swallow text inside of angle brackets. Did you mean you tried `<author>\<title>`?

If so, that should do what I believe you're trying to accomplish. Each book from the same author will be placed inside their own folder, and those folders will be placed inside the same author's folder.

Here's an example using the Harry Potter series:

```
J.K. Rowling\
|- Harry Potter and the Sorcerer's Stone, Book 1\
|  |- Harry Potter and the Sorcerer's Stone, Book.m4b
|  |- Harry Potter and the Sorcerer's Stone, Book.jpg
|- Harry Potter and the Chamber of Secrets, Book 2
|  |- Harry Potter and the Chamber of Secrets, Book 2.m4b
|  |- Harry Potter and the Chamber of Secrets, Book 2.jpg
```

Is that what you're looking for?

**Comment by rmcrackan** [comment]
This was her original text:

```
<author>
\<title> - <series> - <language short> [<id>]
```

**Comment by Mbucari** [comment]
> I'm allowed to edit, which let's me see what was actually typed

Clever boy!

**Comment by KatCountryGirl** [comment]
Thank you guys. Additionally If I wanted to redownload books already done how would I do so? I see there is an option to replace existing files but how to redownload books already downloaded? Or is there a way to rename/organize them in this way? Is there a program you know of or a way of using the naming scheme for files acquired outside libation? Im aware of OpenAudible but it just tags the files it doesnt rename the files. Thank you so much again!

**Comment by CLHatch** [comment]
> Hello first off let me say thank you I LOVE your program.
> 
> I would love the option to rename files by author folder then by book in its own folder inside the author folder. Right now I can do this this author/book files but if there are multiple books they are all just in the author folder. And save the image to that internal book folder.
> 
> I did try to add
> 
> \<title> - - []
> Also the option to use the renamer for files I got outside of libation would be amazing as I have yet to find a decent program to rename my acquired files like yours does. Thanks again!

You might be interested to know that you can go a folder further with renaming patterns, if you want.  I use `Author/Series/Book/Files`  for books that are in a series, and append the book number in the series to the "Book" folder name.

**Comment by KatCountryGirl** [comment]
> @KatCountryGirl I'm glad you're enjoying it!
> 
> > I did try to add
> > <title> - - []
> 
> Github likes to swallow text inside of angle brackets. Did you mean you tried `<author>\<title>`?
> 
> If so, that should do what I believe you're trying to accomplish. Each book from the same author will be placed inside their own folder, and those folders will be placed inside the same author's folder.
> 
> Here's an example using the Harry Potter series:
> 
> ```
> J.K. Rowling\
> |- Harry Potter and the Sorcerer's Stone, Book 1\
> |  |- Harry Potter and the Sorcerer's Stone, Book.m4b
> |  |- Harry Potter and the Sorcerer's Stone, Book.jpg
> |- Harry Potter and the Chamber of Secrets, Book 2
> |  |- Harry Potter and the Chamber of Secrets, Book 2.m4b
> |  |- Harry Potter and the Chamber of Secrets, Book 2.jpg
> ```
> 
> Is that what you're looking for?

exactly! Ill try to tonight when my 24hr api ban runs out :D

**Comment by KatCountryGirl** [comment]
> > Hello first off let me say thank you I LOVE your program.
> > I would love the option to rename files by author folder then by book in its own folder inside the author folder. Right now I can do this this author/book files but if there are multiple books they are all just in the author folder. And save the image to that internal book folder.
> > I did try to add
> > <title> - - []
> > Also the option to use the renamer for files I got outside of libation would be amazing as I have yet to find a decent program to rename my acquired files like yours does. Thanks again!
> 
> You might be interested to know that you can go a folder further with renaming patterns, if you want. I use `Author/Series/Book/Files` for books that are in a series, and append the book number in the series to the "Book" folder name.

Thank you so much thats nice to know. Just needed the "\"

**Comment by KatCountryGirl** [comment]
so I just found the slash commands button :D I did this, fingers crossed.

File Naming
/<author>\<title>\<series>
File Naming 
/<title> - <language short> [<id>]

**Comment by KatCountryGirl** [comment]
so I just found the slash commands button :D I did this, fingers crossed.

File Naming
`<author>\<title>\<series>`
File Naming 
`<title> - <language short> [<id>]`

**Comment by rmcrackan** [comment]
Best of luck! Let us know how it goes.

I'm curious about your ban. How did you find out it was a 24 hour ban? Any time I've suspected this in the past, it was only a guess. Audible was very hush hush about it.

**Comment by CLHatch** [comment]
> > > Hello first off let me say thank you I LOVE your program.
> > > I would love the option to rename files by author folder then by book in its own folder inside the author folder. Right now I can do this this author/book files but if there are multiple books they are all just in the author folder. And save the image to that internal book folder.
> > > I did try to add
> > > <title> - - []
> > > Also the option to use the renamer for files I got outside of libation would be amazing as I have yet to find a decent program to rename my acquired files like yours does. Thanks again!
> > 
> > 
> > You might be interested to know that you can go a folder further with renaming patterns, if you want. I use `Author/Series/Book/Files` for books that are in a series, and append the book number in the series to the "Book" folder name.
> 
> Thank you so much thats nice to know. Just needed the " \ "

Give this a shot, see if you like it, it's what I'm using currently.  Ends up:

Books:  `Account\Books\Author(s)\[Series/## - ](Year) - Title [ASIN] {Narrators(s)}\Short Title [ASIN]`
Podcasts: `Account\Podcasts\[Series\](Date) - Title [ASIN] {Narrators(s)}\Short Title [ASIN]`

Folder Naming: `<account nickname>\<if podcast->Podcasts\<if series-><series>\<-if series>(<pub date>) - <title> [<id>] {<narrator[separator(; )]>}<-if podcast><!if podcast->AudioBooks\<author[separator(; )]>\<if series-><series>\<series#[00.##]> - <-if series>(<year>) - <title> [<id>] {<narrator[separator(; )]>}<-if podcast>`
File Naming: `<title short> [<id>]`

**Comment by KatCountryGirl** [comment]
> Best of luck! Let us know how it goes.
> 
> I'm curious about your ban. How did you find out it was a 24 hour ban? Any time I've suspected this in the past, it was only a guess. Audible was very hush hush about it.

It cuts off at about 130-150 books. Im trying to get the top 100 of every genre during my free trial :P 24 hrs to the minute I can download again. I just load my list up and it downloads nearly 150 books and then the error comes up saying something about not being able to download for 24/48 hrs. Then I just look at file modified time of the last file downloaded and then that's the time I can start again the next day. I just have my task scheduler to start the program every day at that time. I dont have crazy fast internet it usually takes about an hour give or take to get them all. 

I'll let you know tomorrow how it went. Thanks again for your time. And if you have any ideas about a renamer for my other files i would greatly appreciate it.

**Comment by KatCountryGirl** [comment]
> 

Thats a good idea, I'll give it a shot. Thanks.

**Comment by KatCountryGirl** [comment]
> > > > Hello first off let me say thank you I LOVE your program.
> > > > I would love the option to rename files by author folder then by book in its own folder inside the author folder. Right now I can do this this author/book files but if there are multiple books they are all just in the author folder. And save the image to that internal book folder.
> > > > I did try to add
> > > > <title> - - []
> > > > Also the option to use the renamer for files I got outside of libation would be amazing as I have yet to find a decent program to rename my acquired files like yours does. Thanks again!
> > > 
> > > 
> > > You might be interested to know that you can go a folder further with renaming patterns, if you want. I use `Author/Series/Book/Files` for books that are in a series, and append the book number in the series to the "Book" folder name.
> > 
> > 
> > Thank you so much thats nice to know. Just needed the " \ "
> 
> Give this a shot, see if you like it, it's what I'm using currently. Ends up:
> 
> Books: `Account\AudioBooks\Author(s)\[Series\## - ](Year) - Title [ASIN] {Narrators(s)}\Short Title [ASIN]` Podcasts: `Account\Podcasts\[Series\](Date) - Title [ASIN] {Narrators(s)}\Short Title [ASIN]`
> 
> Folder Naming: `<account nickname>\<if podcast->Podcasts\<if series-><series>\<-if series>(<pub date>) - <title> [<id>] {<narrator[separator(; )]>}<-if podcast><!if podcast->AudioBooks\<author[separator(; )]>\<if series-><series>\<series#[00.##]> - <-if series>(<year>) - <title> [<id>] {<narrator[separator(; )]>}<-if podcast>` File Naming: `<title short> [<id>]`
> 
> If you don't have podcasts, and don't care about multiple accounts, the folder naming can be simplified to:
> 
> `<author[separator(; )]>\<if series-><series>\<series#[00.##]> - <-if series>(<year>) - <title> [<id>] {<narrator[separator(; )]>}`

Thats a good idea, I'll give it a shot. Thanks.

**Comment by KatCountryGirl** [comment]
> Best of luck! Let us know how it goes.
> 
> I'm curious about your ban. How did you find out it was a 24 hour ban? Any time I've suspected this in the past, it was only a guess. Audible was very hush hush about it.

Id be willing to pay for a program or plugin that could take my tagged names in the file and rename the actual file to that. That would be amazing!

**Comment by CLHatch** [comment]
> > Best of luck! Let us know how it goes.
> > I'm curious about your ban. How did you find out it was a 24 hour ban? Any time I've suspected this in the past, it was only a guess. Audible was very hush hush about it.
> 
> Id be willing to pay for a program or plugin that could take my tagged names in the file and rename the actual file to that. That would be amazing!

I seem to remember that there was talk of adding something like that to AudioBookShelf.  Not a current feature, but one that they were going to work on.  I don't think it was for the filenames, it was to name the folders they go in.

**Comment by KatCountryGirl** [comment]
Ive tried both audiobookshelf and readarr. Audiobookshelf will tag the files but not rename the actual file and readarr is great but it wont pick up all my books seems to be very picky unless it is already named correctly. 

> 



> > > Best of luck! Let us know how it goes.
> > > I'm curious about your ban. How did you find out it was a 24 hour ban? Any time I've suspected this in the past, it was only a guess. Audible was very hush hush about it.
> > 
> > 
> > Id be willing to pay for a program or plugin that could take my tagged names in the file and rename the actual file to that. That would be amazing!
> 
> I seem to remember that there was talk of adding something like that to AudioBookShelf. Not a current feature, but one that they were going to work on. I don't think it was for the filenames, it was to name the folders they go in.
> 
> You also might be able to do it with Readarr.

Ive tried both audiobookshelf and readarr. Audiobookshelf will tag the files but not rename the actual file and readarr is great for renaming but wont tag the files. It wont pick up all my books seems to be very picky unless it is already named/tagged correctly. Id love something like a mix of readarr and audiobookshelf where it can detect rename and organize the files all at once. When you download books off the internet they might be perfect or it might all just be jumbled letters and numbers. If I only had a few it wouldnt matter but I have quite a bit.

**Comment by CLHatch** [comment]
> > 
> 
> > > > Best of luck! Let us know how it goes.
> > > > I'm curious about your ban. How did you find out it was a 24 hour ban? Any time I've suspected this in the past, it was only a guess. Audible was very hush hush about it.
> > > 
> > > 
> > > Id be willing to pay for a program or plugin that could take my tagged names in the file and rename the actual file to that. That would be amazing!
> > 
> > 
> > I seem to remember that there was talk of adding something like that to AudioBookShelf. Not a current feature, but one that they were going to work on. I don't think it was for the filenames, it was to name the folders they go in.
> > You also might be able to do it with Readarr.
> 
> Ive tried both audiobookshelf and readarr. Audiobookshelf will tag the files but not rename the actual file and readarr is great for renaming but wont tag the files. It wont pick up all my books seems to be very picky unless it is already named/tagged correctly. Id love something like a mix of readarr and audiobookshelf where it can detect rename and organize the files all at once. When you download books off the internet they might be perfect or it might all just be jumbled letters and numbers. If I only had a few it wouldnt matter but I have quite a bit.

Yeah, I wasn't sure if Readarr did it or not.  And I knew AudioBookShelf doesn't currently rename, but I seem to remember it being mentioned by the dev in his Discord server that he was wanting to add renaming.  So you might get your wish at some point.

**Comment by rmcrackan** [comment]
> It cuts off at about 130-150 books. Im trying to get the top 100 of every genre during my free trial

Thanks for confirming. Getting blocked by audible for using Libation is an extremely frequent user concern. The only times I've seen it happen are like this one: tons 'o Plus downloads at once and even then it's only like a 1 day block.

Regarding writing tags, meta data, file renaming, etc -- these are deceptively difficult programs to write. You start with the simple case and it goes great so you power forth with confidence. Then you hit your first snag which isn't too bad, then another which is weirder, and another which is borderline unexplainable. It's the epitome of death by 1,000 cuts as you dig into the weird edge cases and bad decisions that operating systems and audio file makers hacked together over decades. I've been down this road and I couldn't say it any better than Shakespeare did: "O, that way madness lies; let me shun that; No more of that." Inside Libation, there's actually an unholy amount of code dedicated just to getting our file naming correct, especially since we have our powerful custom tagging system.

It's doable and you might find someone to work just with you and your library. If we were to add it as an option in Libation though, we really would have to address *everyone's* edge cases and I'm not down for that.

**Comment by CLHatch** [comment]
I was right, it is something that is being actively worked on in AudioBookShelf.  Someone has created a PR to add the functionality in.  As of the 21st, the main dev said in Discord that he's still looking at it, and that the scanner ABS currently uses is going to have to be re-written to be able to handle scanning custom folders like that.

https://github.com/advplyr/audiobookshelf/pull/1521

**Comment by rmcrackan** [comment]
Sorry, I'm afraid I've lost the plot. Is there a pending request remaining in this ticket?

**Comment by CLHatch** [comment]
> Sorry, I'm afraid I've lost the plot. Is there a pending request remaining in this ticket?

I don't believe so, no. @KatCountryGirl was just needing help with getting her books under an "author" folder, and I shared my renaming pattern (I believe she got it working?), then it shifted to "Would be nice if there were an app rename based on tags".  I just mentioned that that MIGHT be a feature coming to another app.

**Comment by rmcrackan** [comment]
Gracias. I thought that was the case and wanted to be sure.
